File: mliascala/src/main/scala/mlia_nd4j/classifier/kNN/kNN.scala
package mlia_nd4j.classifier.kNN

import org.nd4j.linalg.api.ndarray.INDArray
import org.nd4j.linalg.factory.Nd4j
import org.nd4j.linalg.ops.transforms.Transforms
import org.nd4s.Implicits._

object kNN {
  def classify0(inX: INDArray, dataSet: INDArray, labels: Array[String], k: Int): String = {
    require(k > 0)
    require(dataSet.rows == labels.length)
    require(inX.length() == dataSet.columns())
    val distances = (0 until dataSet.rows()).foldLeft(Nd4j.create(dataSet.rows(), 1)){
      (res, idx) =>{
        res(idx, 0) = dataSet(idx, ->).distance2(inX)
        res
      }
    }
    val sortedDistIndicies: Array[INDArray] = Nd4j.sortWithIndices(distances, 0, true)
    val indecs: Array[Int] = sortedDistIndicies(0).data().asInt().take(k)

    val classCount  = indecs.foldLeft(Map.empty[String, Int]){
      (counter, idx) =>
        val vote = labels(idx)
        counter + (vote -> (counter.getOrElse(vote, 0) + 1))
    }
    classCount.toArray.sortBy(_._2).reverse.headOption.map(_._1).getOrElse("Failure")
  }
  def autoNorm(dataSet: INDArray): (INDArray, INDArray, INDArray) = {

    val minVals = (0 until dataSet.columns()).foldLeft(Nd4j.create(1, dataSet.columns())){
      (array, idx) =>
        Transforms.min(dataSet(->, idx), 1)
    }
    val maxVals = (0 until dataSet.columns()).foldLeft(Nd4j.create(1, dataSet.columns())){
      (array, idx) =>
        Transforms.max(dataSet(->, idx), 1)
    }

    val ranges = maxVals - minVals
    val diff = dataSet - minVals.reshape(minVals.rows(), minVals.columns())//tile1(minVals, m.rows)
    (diff / diff, ranges, minVals)

  }
  def createDataSet  = {
    val dataSet = Nd4j.create(Array[Float](1.0f, 1.1f, 1.0f, 1.0f, 0.0f, 0.0f, 0.0f, 1.1f), Array[Int](4, 2))
    val labels = Array[String]("A", "A", "B", "B")
     (dataSet, labels)
  }

  def main(args: Array[String]): Unit = {
    val (dataSet, labels)  = createDataSet
    val inX = Nd4j.create(Array[Float](0.0f, 1.1f), Array[Int](1, 2))
    printf("classify: %s\n", classify0(inX, dataSet, labels, 1))
  }
}

File: mliajava/src/main/java/classifier/kNN/kNN.java
package classifier.kNN;


import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.indexing.NDArrayIndex;
import org.nd4j.linalg.ops.transforms.Transforms;
import org.nd4j.linalg.util.NDArrayUtil;

import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class kNN {

    public static String classify0(INDArray inX, INDArray dataSet, String[] labels, int k) {
        INDArray distances = Nd4j.create(dataSet.rows(), 1);
        for (int i = 0; i < dataSet.shape()[0]; i++) {
            distances.put(i, 0, dataSet.getRow(i).distance2(inX));
        }
        INDArray[] sortedDistIndicies = Nd4j.sortWithIndices(distances, 0, true);
        INDArray indecs = sortedDistIndicies[0].get(NDArrayIndex.interval(0, k), NDArrayIndex.all());

        Stream<Integer> kIndexs = IntStream.of(NDArrayUtil.toInts(indecs)).boxed();

        Map<String, Integer> counter = kIndexs.reduce(
                new TreeMap<String, Integer>(),
                (c, index) -> {
                    String vote = labels[index];
                    c.put(vote, c.getOrDefault(vote, new Integer(0)) + 1);
                    return c;
                },
                (left, right) -> {
                    left.putAll(right);
                    return left;
                });

        Set<Map.Entry<String, Integer>> rc = new TreeSet<Map.Entry<String, Integer>>((x1, x2) -> {
            return x2.getValue() - x1.getValue() != 0 ? x2.getValue() - x1.getValue() : x2.getKey().compareTo(x1.getKey());
         });

        rc.addAll(counter.entrySet());

        return rc.iterator().next().getKey();
    }

    public static INDArray autoNorm(INDArray dataSet) {

        INDArray minVals = Nd4j.create(1, dataSet.columns());
        for (int i = 0; i < dataSet.columns(); i++) minVals = Transforms.min(dataSet.getColumn(i), 1);
        INDArray maxVals = Nd4j.create(1, dataSet.columns());
        for (int i = 0; i < dataSet.columns(); i++) maxVals = Transforms.max(dataSet.getColumn(i), 1);

        INDArray ranges = maxVals.sub(minVals);

        INDArray res = Nd4j.create(dataSet.rows(), dataSet.columns());
        for (int i = 0; i < dataSet.rows(); i++)
            res.putRow(i, dataSet.getRow(i).sub(minVals).div(ranges));

        INDArray retV = Nd4j.create(3, 1);
        retV.putRow(0, res);
        retV.putRow(1, ranges);
        retV.putRow(2, minVals);
        return retV;

    }
}
